api/envoy/config/common/matcher/v3/matcher.proto

```diff
@@ -16,6 +19,93 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
 // [#protodoc-title: Unified Matcher API]
 
+message MatchTree {
+  message MatchAction {
+    oneof action {
+      option (validate.required) = true;
+
+      // The associated operation should be skipped.
+      bool skip = 1;
+
+      // The specified callback value should be passed to the associated operation.
+      string callback = 2;
+    }
+  }
+
+  message MatchLeaf {
+    // A single match attempt. If this entry matches, the specified action will be performed.
+    message LeafMatcher {
+      oneof matcher {
+        option (validate.required) = true;
+
+        MatchPredicate predicate = 1;
+
+        core.v3.TypedExtensionConfig typed_config = 2;
+      }
+
+      // The action to perform if the predicate evaluates to true.
+      MatchAction action = 3 [(validate.rules).message = {required: true}];
+    }
+
+    // A linear set of match conditions that will be evaluated in order.
+    // Each matcher will be attempted in order until one succeeds.
+    // If none of the matchers match, the no_match_action will be performed if specified.
+    repeated LeafMatcher matchers = 1 [(validate.rules).repeated = {min_items: 1}];
+
+    // Action to perform if none of the above matchers match.
+    MatchAction no_match_action = 3;
+  }
+
+  message SublinearMatcher {
+    message MultimapMatcher {
+      // The header to apply this match to.
+      // The key to apply this matcher for.
+      string key = 1 [(validate.rules).string = {min_len: 1}];
+
+      // The namespace of this key. This is used to distinguish multiple maps
+      // that could be consulted during matching.
+      string key_namespace = 2 [(validate.rules).string = {min_len: 1}];
+
+      // Mapping from exact value matches to the matching subtree to use when
+      // the match applies.
+      // If a header value matches both a prefix and an exact match, the exact
+      // match path will be taken.
+      map<string, MatchTree> exact_matches = 3;
+
+      // Mapping from prefix matches to the matching subtree to use when the match applies.
+      // If a header value matches both a prefix and an exact match, the exact match path will be taken.
+      // If multiple prefixes match, the longest prefix path will be taken.
+      map<string, MatchTree> prefix_matches = 4;
+    }
+
+    oneof matcher {
+      option (validate.required) = true;
+
+      MultimapMatcher multimap_matcher = 1;
+
+      core.v3.TypedExtensionConfig typed_config = 2;
+    }
+
+    // If none of the above matchers match, attempt to match against this subtree.
+    MatchTree no_match_tree = 3;
```

Comment:
Isn't this potentially recursive? What's the rationale for a no_match_tree here vs. just an action? I suppose it's more flexible but it seems potentially confusing and/or recursive.

Reply:
Yeah I changed this to a tree to be more flexible as per @euroelessar suggestion in the doc. This definitely makes it recursive, but so is the tree without this as well. I could see this being useful for modeling a match tree in which you use an exact matcher and you want a more sophisticated matching logic for header values you don't know about, but I could probably be convinced that this makes the tree too flexible and thus harder to reason about.